### ISSUE TITLE
Reported Versions Out of Order After Prerelease Tag

### ISSUE THREAD

**Issue by WalkerCodeRanger** [GitTools/GitVersion#1844]
With the following `GitVersion.yml`
```yaml
mode: ContinuousDeployment
branches:
  master:
    tag: dev
```
The table below shows 4 commits on the master branch, oldest to newest. The middle column indicates what tags have been created in git (`git tag something`). The third column shows the version GitVersion assigns to them.

Commit #| Git Tag | GitVersion
--- | --- | ---
1 | v2.0.0 | 2.0.0
2 | | 2.0.1-dev.1
3| v2.1.0-rc | 2.1.0-rc
4| | 2.1.0-dev.1

This all seems correct up until commit 4. Notice that according to SemVer 2.0.0 < 2.0.1-dev.1 < 2.1.0-rc, _but_  2.1.0-rc **>** 2.1.0-dev1.

After a release git tag, GitVersion increments the version number, but after a prerelease git tag, it doesn't. That makes some sense if `tag: ''` is set and the new version wil be `-rc.X`, but doesn't make sense in this case.

I think that if the branch tag setting would cause the current version to be before the previous tag, then the increment should happen.

**Comment by asbjornu** [comment]
Which version of GitVersion are you using? Are you able to reproduce the error in a `RepositoryFixture` test akin to the following?

https://github.com/GitTools/GitVersion/blob/92c11a1ddc7ce43490084e94bff4519e60a92d0d/src/GitVersionCore.Tests/IntegrationTests/MasterScenarios.cs#L35-L56

**Comment by WalkerCodeRanger** [comment]
I'm using version 5.0.1+Branch.master.Sha.c71b8fc9f6d7b6adffe82fef588e717beb864e91

I've created a failing unit test that shows the issue.

https://github.com/WalkerCodeRanger/GitVersion/blob/fdbe5b4fed22b81c6dc295b5e2c7e5ec9c36211e/src/GitVersionCore.Tests/IntegrationTests/VersionBumpingScenarios.cs#L38-L63

Should I submit a pull request with that code?

**Comment by asbjornu** [comment]
A PR would be great. I you would dare attempt to figure out why GitVersion is acting up, that would be even better. 😃

**Comment by WalkerCodeRanger** [comment]
Debugging the unit test in my PR #1845 to see why this is happening...

`NextVersionCalculator.FindVersion()` gets the base version and then tries to increment it and correct the prerelease tag:
https://github.com/GitTools/GitVersion/blob/92c11a1ddc7ce43490084e94bff4519e60a92d0d/src/GitVersionCore/VersionCalculation/NextVersionCalculator.cs#L54-L64

`PerformIncrement()` ends up calling `SemanticVersion.IncrementVersion()` on a pre-release version (the base version coming from the tag) with `VersionField.Patch`.  But that checks for pre-release:
https://github.com/GitTools/GitVersion/blob/92c11a1ddc7ce43490084e94bff4519e60a92d0d/src/GitVersionCore/SemanticVersion.cs#L291
and then only increments the pre-release tag number if there is one:
https://github.com/GitTools/GitVersion/blob/92c11a1ddc7ce43490084e94bff4519e60a92d0d/src/GitVersionCore/SemanticVersion.cs#L315-L319
To me, that seems wrong. I asked it to increment the patch version and it didn't. Changing `SemanticVersion.IncrementVersion()` to not treat pre-release special, but to go ahead and increment (but possibly remove the pre-release tag) makes more sense. But doing that causes many more unit tests to break because the whole system depends on the current behavior.

I guess `NextVersionCalculator.UpdatePreReleaseTag()` could patch up the version number, but I don't understand things well enough to know what change to make there. Alternatively, `NextVersionCalculator.FindVersion()` could use `preReleaseTagDoesNotMatchConfiguration` to remove the prerelease tag and try to increment again, but that seems like a hack.

Is there any guidance you could offer on how to fix this?

**Comment by asbjornu** [comment]
Yes, now that you're starting to dig into this, I remember this being a complicated hot mess. Perhaps @arturcic remembers when we discussed this last and what the result of that discussion was? My memory is not my strongest asset. 😅

**Comment by asbjornu** [comment]
Will the fix implemented in #2003 help with this problem in any way?
